code/game/machinery/doors/firedoor.dm

```diff
@@ -21,14 +22,62 @@
 	assemblytype = /obj/structure/firelock_frame
 	armor = list(MELEE = 10, BULLET = 30, LASER = 20, ENERGY = 20, BOMB = 30, BIO = 100, RAD = 100, FIRE = 95, ACID = 70)
 	interaction_flags_machine = INTERACT_MACHINE_WIRES_IF_OPEN | INTERACT_MACHINE_ALLOW_SILICON | INTERACT_MACHINE_OPEN_SILICON | INTERACT_MACHINE_REQUIRES_SILICON | INTERACT_MACHINE_OPEN
-	var/nextstate = null
+
+	COOLDOWN_DECLARE(detect_cooldown)
+
+	///Trick to get the glowing overlay visible from a distance
+	luminosity = 1
+	///X offset for the overlay lights, so that they line up with the thin border firelocks
+	var/light_xoffset = 0
+	///Y offset for the overlay lights, so that they line up with the thin border firelocks
+	var/light_yoffset = 0
+
 	var/boltslocked = TRUE
+	///List of areas we handle. See CalculateAffectingAreas()
 	var/list/affecting_areas
+	///For the few times we affect only the area we're actually in. Set during Init. If we get moved, we don't update, but this is consistant with fire alarms and also kinda funny so call it intentional.
+	var/area/my_area
+	///Tracks if the firelock is being held open by a crowbar. If so, we don't close until they walk away
 	var/being_held_open = FALSE
+	///Type of alarm when active. See code/defines/firealarm.dm for the list. This var being null means there is no alarm.
+	var/alarm_type = null
+	///The merger_id and merger_typecache variables are used to make rows of firelocks activate at the same time.
+	var/merger_id = "firelocks"
+	var/merger_typecache
+	///Overlay object for the warning lights. This and some plane settings allows the lights to glow in the dark.
+	var/mutable_appearance/warn_lights
 
 /obj/machinery/door/firedoor/Initialize(mapload)
 	. = ..()
+	AddElement(/datum/element/atmos_sensitive, mapload)
 	CalculateAffectingAreas()
+	my_area = get_area(src)
+	var/static/list/loc_connections = list(
+		COMSIG_TURF_EXPOSE = .proc/check_atmos,
+	)
+	AddElement(/datum/element/connect_loc, loc_connections)
+	if(!merger_typecache)
+		merger_typecache = typecacheof(/obj/machinery/door/firedoor)
+
+	check_atmos()
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/**
+ * Sets the offset for the warning lights.
+ *
+ * Used for special firelocks with light overlays that don't line up to their sprite.
+ */
+/obj/machinery/door/firedoor/proc/set_offset()
```

Comment:
update_offset maybe? set implies it accepts an argument to set the thing to

Reply:
Changed to `adjust_lights_starting_offset()`.